flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java

```diff
@@ -270,7 +271,13 @@
 		 * ALIGNED_EXACTLY_ONCE_CHECKPOINT_BARRIER indicates that this buffer represents a
 		 * serialized checkpoint barrier of aligned exactly-once checkpoint mode.
 		 */
-		ALIGNED_EXACTLY_ONCE_CHECKPOINT_BARRIER(false, true, true, false);
+		ALIGNED_EXACTLY_ONCE_CHECKPOINT_BARRIER(false, true, true, false),
+
+		/**
+		 * Indicates that this subpartition state is fully recovered (emitted).
+		 * Further data can be consumed after unblocking.
+		 */
+		RECOVERY_COMPLETION(false, true, true, false);
```

Comment:
We should probably generalize ALIGNED_EXACTLY_ONCE_CHECKPOINT_BARRIER + RECOVERY_COMPLETION to some BLOCKING_EVENT type.

Reply:
I'd keep these events separate as long as they mean different things.
Even if they are treated the same way today, doesn't mean they will in the future.